pkg/operator/csr/certificate.go

```diff
@@ -0,0 +1,135 @@
+			return false, nil
+		}
+	}
+
+	if subject == nil {
+		return true, nil
+	}
+
+	// check subject of certificates
+	for _, cert := range certs {
+		if cert.Subject.CommonName != subject.CommonName {
+			continue
+		}
+		return true, nil
+	}
+
+	klog.V(4).Infof("Certificate is not issued for subject (cn=%s)", subject.CommonName)
+	return false, nil
+}
+
+// getCertValidityPeriod returns the validity period of the client certificate in the secret
+func getCertValidityPeriod(secret *corev1.Secret) (*time.Time, *time.Time, error) {
+	if secret.Data == nil {
+		return nil, nil, fmt.Errorf("no client certificate found in secret %q", secret.Namespace+"/"+secret.Name)
+	}
+
+	certData, ok := secret.Data[TLSCertFile]
+	if !ok {
+		return nil, nil, fmt.Errorf("no client certificate found in secret %q", secret.Namespace+"/"+secret.Name)
+	}
+
+	certs, err := certutil.ParseCertsPEM(certData)
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to parse TLS certificates: %w", err)
+	}
+
+	if len(certs) == 0 {
+		return nil, nil, errors.New("No cert found in certificate")
+	}
+
+	// find out the validity period for all certs in the certificate chain
+	var notBefore, notAfter *time.Time
+	for index, cert := range certs {
+		if index == 0 {
+			notBefore = &cert.NotBefore
+			notAfter = &cert.NotAfter
+			continue
+		}
+
+		if notBefore.Before(cert.NotBefore) {
+			notBefore = &cert.NotBefore
+		}
+
+		if notAfter.After(cert.NotAfter) {
+			notAfter = &cert.NotAfter
+		}
+	}
+
+	return notBefore, notAfter, nil
+}
+
+// BuildKubeconfig builds a kubeconfig based on a rest config template with a cert/key pair
+func BuildKubeconfig(clientConfig *restclient.Config, certPath, keyPath string) clientcmdapi.Config {
+	// Build kubeconfig.
+	kubeconfig := clientcmdapi.Config{
+		// Define a cluster stanza based on the bootstrap kubeconfig.
+		Clusters: map[string]*clientcmdapi.Cluster{"default-cluster": {
+			Server:                   clientConfig.Host,
+			InsecureSkipTLSVerify:    false,
+			CertificateAuthorityData: clientConfig.CAData,
+		}},
+		// Define auth based on the obtained client cert.
+		AuthInfos: map[string]*clientcmdapi.AuthInfo{"default-auth": {
+			ClientCertificate: certPath,
+			ClientKey:         keyPath,
+		}},
+		// Define a context that connects the auth info and cluster, and set it as the default
+		Contexts: map[string]*clientcmdapi.Context{"default-context": {
+			Cluster:   "default-cluster",
+			AuthInfo:  "default-auth",
+			Namespace: "configuration",
+		}},
+		CurrentContext: "default-context",
+	}
+
+	return kubeconfig
+}
+
+// isCSRApproved returns true if the given csr has been approved
+func isCSRApproved(csr *certificatesv1.CertificateSigningRequest) bool {
+	approved := false
+	for _, condition := range csr.Status.Conditions {
+		if condition.Type == certificatesv1.CertificateDenied {
+			return false
+		} else if condition.Type == certificatesv1.CertificateApproved {
+			approved = true
+		}
```

Comment:
No even on either a Denied or Failed certificate request?

`CertificateApproved == true` condition appearing before `(CertificateDenied == true || CertficiateFailed == true)` can still result in this being accepted as an approved CSR. While technically that's an invalid state, I don't feel confident in using a certificate from an invalid CSR.